library/core/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,88 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer with
+/// a different signature, the two signatures must be *ABI-compatible* or else calling the function
+/// via that function pointer is Undefined Behavior. ABI compatibility is a lot stricter than merely
+/// having the same memory layout; for example, even if `i32` and `f32` have the same size and
+/// alignment, they might be passed in different registers and hence not be ABI-compatible.
+///
+/// ABI compatibility as a concern only arises in code that alters the type of function pointers,
+/// which is wildly unsafe. It is significantly more wildly unsafe than even
+/// [`transmute_copy`][mem::transmute_copy], which is about the most unsafe function in this
+/// library. Function calls should use exactly the same signature on the caller and callee side if
+/// at all possible, and only resort to relying on ABI compatibility in exceptional circumstances.
+/// But assuming such circumstances, what are the rules?
+///
+/// For two signatures to be considered *ABI-compatible*, they must use a compatible ABI string (as
+/// declared via `extern "ABI string"`), must take the same number of arguments, and the individual
+/// argument types and the return types must be ABI-compatible.
+///
+/// The ABI strings are guaranteed to be compatible if they are the same, or if the caller ABI string is
+/// `$X-unwind` and the callee ABI string is `$X`.
```

Comment:
I should probably say that if the caller uses a non-unwinding ABI, but the callee unwinds anyway, then that is clearly UB.

Reply:
Ah wait no that case is already not allowed since only the caller may have an extra `-unwind`.